lib/tasks/db/update_annotation_line.rake

```diff
@@ -0,0 +1,107 @@
+# frozen_string_literal: true
+namespace :db do
+  task update_annotation_line: :environment do
+    Rails.application.eager_load!
+
+    require 'action_view'
+    require 'action_view/helpers'
+
+    helper = HelperContext.new
+
+    start_date_input, end_date_input = ENV['START_DATE'], ENV['END_DATE']
+    unless start_date_input && end_date_input
+      puts 'Please provide START_DATE and END_DATE to execute this script'
+      exit
+    end
+
+    start_date = Date.parse(start_date_input)
+    end_date = Date.parse(end_date_input)
+
+    if start_date > end_date
+      puts 'START_DATE should be earlier than END_DATE'
+      exit
+    end
+
+    start_time = Time.now
+
+    ActsAsTenant.without_tenant do
+      conn = ActiveRecord::Base.connection
+      annotations_with_file_content = conn.exec_query(<<-SQL)
+        SELECT annotations.*, files.content, questions.actable_id
+        FROM course_assessment_answer_programming_file_annotations AS annotations
+        JOIN course_assessment_answer_programming_files AS files
+          ON annotations.file_id = files.id
+        JOIN course_assessment_answers AS answers
+          ON files.answer_id = answers.actable_id
+          AND answers.actable_type = 'Course::Assessment::Answer::Programming'
+          AND CAST(answers.created_at AS DATE) BETWEEN '#{start_date}' AND '#{end_date}'
+        JOIN course_assessment_questions AS questions
+          ON answers.question_id = questions.id
+      SQL
+
+      end_time = Time.now
+      puts "Extract all annotations with file content = #{end_time - start_time} seconds"
+
+      annotations_grouped = annotations_with_file_content.group_by do |row|
+        [row['content'], row['actable_id'], row['file_id']]
+      end
+
+      start_time = Time.now
+
+      annotations_grouped.map do |key, values|
+        content, actable_id, file_id = key
+        programming_question = Course::Assessment::Question::Programming.find(actable_id)
+        language = programming_question.language
+
+        lines = values.map { |value| value['line'] }
+        highlighted_code = helper.highlight_code_block(content, language)
+
+        if !annotations_present_on_line_one_and_two(lines) && lines_are_added(highlighted_code)
+          decrement_annotation_line_number(file_id)
+        end
+      end
+
+      end_time = Time.now
+      puts "Update annotation lines = #{end_time - start_time} seconds"
```

Comment:
Can report how long this takes for roughly for 6 months range 2022-07-01 to 2023-01-01? To get a rough estimate.

Reply:
I have previously applied the decrement of the annotation line for that period, so instead I try with another 6 months range (`2021-07-01` to `2022-01-01`), and here are the time taken..

- extract the data from the DB (executing SQL query) -> 8.1 seconds
- grouping those annotations based on `file_id` -> 0.06 seconds
- applying highlighter and update annotation properly -> 161.31 seconds